### PR TITLE
Use behavior api for eth protocol

### PR DESCRIPTION
**Pull request by pipermerriam**
builds on #1150 

### What was wrong?

We need to get rid of `ETHPeer` and `LESPeer`.  These both have functionality that can be replaced by behaviors.

### How was it fixed?

Create `LESAPI` and `ETHAPI` and move the handling of `NewBlock/Announce` messages into these.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

#### Cute Animal Picture

17248b8aad8eb5c834f2206acfb53fd2--corgi-costume-cute-dog-costumes

### CODE REVIEW

File: p2p/behaviors.py [hunk 1]
```diff
@@ -1,152 +1,72 @@
...
-            raise Exception("Reentrance!")
-        self._connection = connection
-        yield
+        # once the logic is bound to the connection we enter it's context.
+        async with self.logic(connection):
+            yield
 
 
-class Application(BehaviorAPI):
+class BehaviourGroup(BehaviorAPI):
```
Comment (pipermerriam) [inline p2p/behaviors.py]: I think I can remove this class and just inline this logic in the one place it is used.

File: p2p/connection.py [hunk 2]
```diff
@@ -244,6 +245,9 @@ def get_p2p_receipt(self) -> DevP2PReceipt:
     #
     # Protocol APIS
     #
+    def has_protocol(self, protocol_identifier: Union[ProtocolAPI, Type[ProtocolAPI]]) -> bool:
+        return self._multiplexer.has_protocol(protocol_identifier)
```
Comment (pipermerriam) [inline p2p/connection.py]: should be pulled out into separate PR

File: p2p/logic.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        self.is_bound = True
+
+    def as_behavior(self, qualifier: QualifierFn = None) -> BehaviorAPI:
+        if qualifier is None:
+            if self.qualifier is None:
+                raise TypeError("No qualifier provided or found on class")
+            qualifier = self.qualifier
+        return Behavior(qualifier, self)
+
+    def bind(self, connection: ConnectionAPI) -> None:
```
Comment (pipermerriam) [inline p2p/logic.py]: I think we can get rid of the `bind` API since `connection` is passed into the `LogicAPI.__call__` and we'll leave it up to the `LogicAPI` decide whether it wants to hold a reference to the `connection` or not.

File: p2p/logic.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    async def __call__(self, connection: ConnectionAPI) -> AsyncIterator[None]:
+        with connection.add_command_handler(self._command_type, self.handle):
+            yield
+
+    @abstractmethod
+    async def handle(self, connection: ConnectionAPI, msg: Payload) -> None:
+        ...
+
+
+def command_handler(command_type: Type[CommandAPI],
```
Comment (pipermerriam) [inline p2p/logic.py]: I think I can remove this for now since it isn't used.

File: p2p/logic.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    When applied an `Application` registers itself with the `ConnectionAPI`
+    under the defined `name`.
+    """
+    name: str
+    _children: Tuple[Tuple[LogicAPI, QualifierFn], ...] = ()
+
+    def add_child_logic(self,
+                        child: LogicAPI,
+                        qualifier: QualifierFn = None) -> Tuple[BehaviorAPI, ...]:
+        self._children += ((child, qualifier),)
```
Comment (pipermerriam) [inline p2p/logic.py]: I think that instead of this function taking `child, qualifier` I'll just have it take a `BehaviorAPI` so that it becomes `self.add_child_behavior(some_logic.as_behavior())`

File: p2p/p2p_api.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    client_version_string: str
+    safe_client_version_string: str
+    logger = logging.getLogger('p2p.p2p_api.P2PAPI')
+
+    local_disconnect_reason: DisconnectReason = None
+
+    def __init__(self) -> None:
+        self.add_child_logic(PongWhenPinged())
+        self._disconnect_handler = CancelOnDisconnect()
+        self.add_child_logic(self._disconnect_handler)
```
Comment (pipermerriam) [inline p2p/p2p_api.py]: Worth pointing out that the only place that this is the only code in this class that deals with the `Behavior/Logic/Qualifier` API.  Everything else is just business logic for the `p2p` protocol.

File: tests/p2p/test_p2p_api.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+@pytest.fixture
+def bob(alice_and_bob):
+    _, bob = alice_and_bob
+    return bob
+
+
+@pytest.mark.asyncio
+async def test_p2p_api_properties(bob, alice):
+    async with P2PAPI().as_behavior().apply(alice):
```
Comment (pipermerriam) [inline tests/p2p/test_p2p_api.py]: This is a bit ugly but it can probably be cleaned up.

File: trinity/protocol/common/api.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+from p2p.qualifiers import HasProtocol
+
+from trinity.protocol.eth.api import ETHAPI
+from trinity.protocol.eth.proto import ETHProtocol
+from trinity.protocol.les.api import LESAPI
+from trinity.protocol.les.proto import LESProtocol, LESProtocolV2
+
+from .abc import ChainInfoAPI, HeadInfoAPI
+
+AnyETHLES = HasProtocol(ETHProtocol) | HasProtocol(LESProtocolV2) | HasProtocol(LESProtocol)
```
Comment (pipermerriam) [inline trinity/protocol/common/api.py]: Here's a good example of the qualifier API in use.

File: trinity/protocol/eth/api.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+class ETHAPI(Application):
+    name = 'eth'
+    qualifier = HasProtocol(ETHProtocol)
+
+    head_info: HeadInfoTracker
+
+    def __init__(self) -> None:
+        self.head_info = HeadInfoTracker()
+        self.add_child_logic(self.head_info, HasCommand(NewBlock))
```
Comment (pipermerriam) [inline trinity/protocol/eth/api.py]: Again, this and the `qualifier` line above are the only places where we're implementing stuff specific to the `behavior/logic/qualifier` APIs, everything else is just business logic.

File: trinity/protocol/eth/handshaker.py [hunk 10]
```diff
@@ -22,6 +25,26 @@ def __init__(self, protocol: ETHProtocol, handshake_params: ETHHandshakeParams)
...
+    def network_id(self) -> int:
+        return self.handshake_params.network_id
+
+    @cached_property
+    def total_difficulty(self) -> int:
+        return self.handshake_params.total_difficulty
+
+    @cached_property
+    def version(self) -> int:
+        return self.handshake_params.version
```
Comment (pipermerriam) [inline trinity/protocol/eth/handshaker.py]: These can be extracted into a stand-alone PR

File: trinity/protocol/les/api.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    # HeadInfoAPI
+    #
+    _head_td: int = None
+    _head_hash: Hash32
+    _head_number: BlockNumber
+
+    @property
+    def head_td(self) -> int:
+        if self._head_td is None:
+            les_receipt = self.connection.get_receipt_by_type(LESHandshakeReceipt)
```
Comment (pipermerriam) [inline trinity/protocol/les/api.py]: This can be simplified by adding a computed `self.les_receipt` property to reduce the repetition of this line.

File: trinity/protocol/les/handshaker.py [hunk 12]
```diff
@@ -36,6 +39,26 @@ def __init__(self,
...
+    def head_hash(self) -> Hash32:
+        return self.handshake_params.head_hash
+
+    @cached_property
+    def head_number(self) -> BlockNumber:
+        return self.handshake_params.head_number
+
+    @cached_property
+    def genesis_hash(self) -> Hash32:
+        return self.handshake_params.genesis_hash
```
Comment (pipermerriam) [inline trinity/protocol/les/handshaker.py]: Can be pulled out to a stand-alone PR

File: trinity/protocol/les/proto.py [hunk 13]
```diff
@@ -136,6 +137,23 @@ def send_handshake(self, handshake_params: LESHandshakeParams) -> None:
...
+        data = {
+            'head_hash': head_hash,
+            'head_number': head_number,
+            'head_td': head_td,
+            'reorg_depth': reorg_depth,
+            'params': params,
+        }
+        cmd = Announce(self.cmd_id_offset, self.snappy_support)
+        self.logger.debug("Sending LES/Announce msg: %s", data)
+        self.transport.send(*cmd.encode(data))
```
Comment (pipermerriam) [inline trinity/protocol/les/proto.py]: stand-alone PR

File: trinity/tools/factories.py [hunk 14]
```diff
@@ -184,7 +184,7 @@ class Meta:
     network_id = MAINNET_NETWORK_ID
     head_td = GENESIS_DIFFICULTY
     head_hash = MAINNET_GENESIS_HASH
-    head_num = GENESIS_BLOCK_NUMBER
+    head_number = GENESIS_BLOCK_NUMBER
```
Comment (pipermerriam) [inline trinity/tools/factories.py]: standalone PR

**Comment by pipermerriam** [comment]
Well, I borked this branches history (will fix) but wanted to drop a note that I've been running this version of the branch for a few hours and nothing seems out of place/broken because of it.

File: tests/core/p2p-proto/test_eth_api.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+
+@pytest.mark.asyncio
+async def test_eth_api_properties(alice):
+    assert alice.connection.has_logic(ETHAPI.name)
+    eth_api = alice.connection.get_logic(ETHAPI.name, ETHAPI)
+
+    assert eth_api is alice.eth_api
+
+    eth_receipt = alice.connection.get_receipt_by_type(ETHHandshakeReceipt)
```
Comment (carver) [inline tests/core/p2p-proto/test_eth_api.py]: That's a neat API.

File: tests/core/p2p-proto/test_les_api.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+@pytest.fixture
+async def alice_and_bob(alice_headerdb, bob_chain):
+    pair_factory = LESV2PeerPairFactory(
+        alice_client_version='alice',
+        alice_peer_context=ChainContextFactory(headerdb=alice_headerdb),
+        bob_client_version='bob',
+        bob_peer_context=ChainContextFactory(headerdb=AsyncHeaderDB(bob_chain.headerdb.db)),
+    )
+    async with pair_factory as (alice, bob):
+        yield alice, bob
```
Comment (carver) [inline tests/core/p2p-proto/test_les_api.py]: It took me a little while to get bob and alice straight in my head, then I *think* I realized what it is. `alice` here, is the `peer` object that would be on "Alice's node". From Alice's perspective, the peer object represents Bob.

Does that sound right? If so, I'm not really sure what to do about it to make that clearer. This might have helped:
```suggestion
        yield alices_bob, bobs_alice
```

Or maybe `to_bob, to_alice`, or even just reversing them. :man_shrugging:

File: trinity/protocol/common/peer.py [hunk 17]
```diff
@@ -59,11 +58,19 @@ class BaseChainPeer(BasePeer):
...
-    genesis_hash: Hash32 = None
+    @cached_property
+    def head_info(self) -> HeadInfoAPI:
+        return self.connection.get_logic(HeadInfo.name, HeadInfo)
+
+    @cached_property
+    def chain_info(self) -> ChainInfoAPI:
+        return self.connection.get_logic(ChainInfo.name, ChainInfo)
+
+    def get_behaviors(self) -> Tuple[BehaviorAPI, ...]:
```
Comment (carver) [inline trinity/protocol/common/peer.py]: Is this `get_behaviors()` API for future use, or did I miss it somewhere in this PR?

File: trinity/sync/light/service.py [hunk 18]
```diff
@@ -296,7 +296,7 @@ def callback(value: Payload) -> None:
             # our best peer doesn't have the header we want, there are no eligible peers.
             raise NoEligiblePeers(f"Our best peer does not have the header {block_hash}")
 
-        head_number = peer.head_number
+        head_number = peer.les_api.head_info.head_number
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply trinity/sync/light/service.py]: This can actually just be updated to be `head_number = peer.head_info.head_number`.  I wrote this code before I wrote the stuff in `trinity.protocol.common.api`

File: tests/core/p2p-proto/test_eth_api.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+@pytest.fixture
+def bob(alice_and_bob):
+    _, bob = alice_and_bob
+    return bob
+
+
+@pytest.mark.asyncio
+async def test_eth_api_properties(alice):
+    assert alice.connection.has_logic(ETHAPI.name)
+    eth_api = alice.connection.get_logic(ETHAPI.name, ETHAPI)
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/core/p2p-proto/test_eth_api.py]: I'll submit to YAGNI.  I have been expecting us to register multiple instances of the same API but it'll be an easy API change to make later of we need that functionality.

File: tests/core/p2p-proto/test_les_api.py (hunk 16 above)
Comment (pipermerriam) [reply tests/core/p2p-proto/test_les_api.py]: Probably not the answer you're looking for but it kind of doesn't matter who's who.  Just that one is one, and the other is the other....  `alice/bob` are easier to keep mentally separated than `peer_a/peer_b`, etc.

File: trinity/sync/light/service.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (carver) [reply trinity/sync/light/service.py]: Ah great, yeah that solves it.

File: tests/core/p2p-proto/test_les_api.py (hunk 16 above)
Comment (carver) [reply tests/core/p2p-proto/test_les_api.py]: Let me see if I can point to a specific place it tripped me up. (and why I think it matters)

File: tests/core/p2p-proto/test_eth_api.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+    assert eth_api.head_info.head_hash == eth_receipt.head_hash
+    assert eth_api.head_info.head_td == eth_receipt.total_difficulty
+    assert not hasattr(eth_api, 'head_number')
+
+
+@pytest.mark.asyncio
+async def test_eth_api_head_info_updates_with_newblock(alice, bob, bob_chain):
+    # mine two blocks on bob's chain
+    bob_chain = build(
+        bob_chain,
```
Comment (carver) [inline tests/core/p2p-proto/test_eth_api.py]: Here, we've done something that affects only one peer: Bob. Now we need to know which peer is "bob's node talking to alice" and which one is "alice's node talking to bob"

Comment (pipermerriam) [reply tests/core/p2p-proto/test_eth_api.py]: Ah, ok, I see.  I'll re-work this to be more clear.

File: tests/core/p2p-proto/test_eth_api.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+    assert eth_api.head_info.head_hash == bob_genesis.hash
+    assert eth_api.head_info.head_td == bob_genesis.difficulty
+    assert not hasattr(eth_api.head_info, 'head_number')
+
+    eth_proto = bob.connection.get_protocol_by_type(ETHProtocol)
+    head = bob_chain.get_canonical_head()
+    assert head.block_number == 2
+    head_block = bob_chain.get_block_by_header(head)
+    total_difficulty = bob_chain.headerdb.get_score(head.hash)
+    eth_proto.send_new_block(
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/core/p2p-proto/test_eth_api.py]: holding off on this change as it'll get cleaned up in #1189

File: tests/core/p2p-proto/test_eth_api.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/core/p2p-proto/test_eth_api.py]: Punting on cleaning this up as to not mix changes to the base API with these ETH/LES protocol changes.
